src/main/java/me/jellysquid/mods/hydrogen/common/jvm/ClassDefineTool.java

```diff
@@ -5,28 +5,54 @@
 import org.apache.logging.log4j.Logger;
 
 import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.net.URL;
 
 public class ClassDefineTool {
     private static final Method DEFINE;
     private static final Logger LOGGER = LogManager.getLogger("Hydrogen-Reflect");
+    private static final boolean IS_JAVA_9 = isJava9OrGreater();
 
     static {
-        LOGGER.warn("Trying to reflect-hack into the class-loader so we can define our own classes in an unrestricted loader...");
-        LOGGER.warn("** The JVM might warn you with something nasty! **");
+        if (IS_JAVA_9) {
+            try {
+                DEFINE = MethodHandles.Lookup.class.getMethod("defineClass", byte[].class);
+            } catch (NoSuchMethodException e) {
+                throw new RuntimeException("Could not obtain reference to MethodHandles.Lookup.defineClass(byte[])", e);
+            }
+        } else {
+            LOGGER.warn("Trying to reflect-hack into the class-loader so we can define our own classes in an unrestricted loader...");
+            LOGGER.warn("** The JVM might warn you with something nasty! **");
+            try {
+                DEFINE = ClassLoader.class.getDeclaredMethod("defineClass", byte[].class, int.class, int.class);
+                DEFINE.setAccessible(true);
+            } catch (NoSuchMethodException e) {
+                throw new RuntimeException("Could not obtain reference to ClassLoader.defineClass(byte[], int, int)", e);
+            } catch (Throwable e) {
+                throw new RuntimeException("Could not expose method", e);
+            }
+        }
+    }
 
+    /**
+     * Checks for the presence of required Java 9 method {@link java.lang.invoke.MethodHandles#privateLookupIn(Class, MethodHandles.Lookup)}.
+     *
+     * @return true if the methods exist (on Java 9 or higher), false otherwise
+     */
+    private static boolean isJava9OrGreater() {
         try {
-            DEFINE = ClassLoader.class.getDeclaredMethod("defineClass", byte[].class, int.class, int.class);
-            DEFINE.setAccessible(true);
-        } catch (NoSuchMethodException e) {
-            throw new RuntimeException("Could not obtain reference to ClassLoader.defineClass(byte[], int, int)", e);
-        } catch (Throwable e) {
-            throw new RuntimeException("Could not expose method", e);
+            Class.forName("java.lang.invoke.MethodHandles").getDeclaredMethod("privateLookupIn", Class.class, MethodHandles.Lookup.class);
```

Comment:
Hmm, should you really check this, or should you check if `Runtime.version()` method exists?

Also if you do check like this, I suggest you use method handle lookup to find this method instead. You can also just use MethodHandles.class because it already exists in 8

Reply:
I check for `privateLookupIn` because that's the method we need in `getTarget` later in the code. It also indicates whether Java >= 9 is being used, but that's just a convenient name for what we actually need: `privateLookupIn` and `MethodHandles.Lookup::defineClass`